### PR TITLE
SIMSBIOHUB-333: Telemetry Table

### PR DESCRIPTION
**Pull request by curtisupshall**
## Links to Jira Tickets

[SIMSBIOHUB-333](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-333)

## Description of Changes

Copied the observations architecture and workflow for managing telemetry and deployments for critters.

## Notes
The file uploading will now be finished in [SIMSBIOHUB-403](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-403) because of the complexity in handling uploaded data 

## Testing Notes

- Navigate to a survey
- Click 'Manage Animals' and add an animal
- Navigate back to survey page and click 'Manage Telemetry'
- Click 'Add Deployment' and fill out the form
- Click 'Add Record' and fill out table row

### CODE REVIEW

File: api/src/services/telemetry-service.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    // step 2 get s3 record for given key
+    const s3Object = await getFileFromS3(submission.key);
+
+    // step 3 parse the file
+    const mediaFile = parseS3File(s3Object);
+
+    // step 4 validate csv
+    if (mediaFile.mimetype !== 'text/csv') {
+      throw new Error(
+        'Failed to process file for importing telemetry. Incorrect file type. Expected CSV received ' +
```
Comment (curtisupshall) [inline api/src/services/telemetry-service.ts]: Nit: should use a template string

File: app/src/components/loading/SkeletonLoaders.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+          position: 'absolute',
+          width: '100%',
+          height: '100%',
+          p: 1,
+          background: grey[100],
+          zIndex: 2
+        }}>
+        <Paper elevation={0} sx={{ overflow: 'hidden' }}>
+          {/* create an array of X items to build multiple skeleton components*/}
+          {Array(props.numberOfLines ?? 10)
```
Comment (curtisupshall) [inline app/src/components/loading/SkeletonLoaders.tsx]: I think Jer wants the usual number of rows in a skeleton layout to be in the neighborhood of 4. @jeznorth

File: app/src/components/loading/SkeletonLoaders.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      </Box>
+    </Fade>
+  );
+};
+
+const GridTableRowSkeleton = (props: IMultipleSkeletonProps) => {
+  return (
+    <Fade in={props.isLoading} timeout={props.timeout ?? 1000}>
+      <Box display="flex" flexDirection="column">
+        {Array(props.numberOfLines ?? 15)
```
Comment (curtisupshall) [inline app/src/components/loading/SkeletonLoaders.tsx]: Same comment re: skeleton layout rows

File: app/src/components/loading/SkeletonLoaders.tsx
Comment (curtisupshall) [inline app/src/components/loading/SkeletonLoaders.tsx:1]: It's probably worth borrowing the latest SkeletonLoader code from the Observations Table. Jer recently made a few changes and simplified these components.

File: app/src/constants/i18n.ts [hunk 4]
```diff
@@ -403,10 +405,32 @@ export const ObservationsTableI18N = {
...
+    `Deleted ${count} ${p(count, 'observation record')} successfully.`
+};
+
+export const TelemetryTableI18N = {
+  removeAllDialogTitle: 'Discard changes?',
+  removeAllDialogText: 'Are you sure you want to discard all your changes? This action cannot be undone.',
+  removeSingleRecordDialogTitle: 'Delete record?',
+  removeSingleRecordDialogText: 'Are you sure you want to delete this record? This action cannot be undone.',
+  removeSingleRecordButtonText: 'Delete Record',
+  removeMultipleRecordsDialogTitle: (count: number) => `Delete ${count} records?`,
```
Comment (curtisupshall) [inline app/src/constants/i18n.ts]: This should use the `pluralize` helper:

```typescript
removeMultipleRecordsDialogTitle: (count: number) => `Delete ${count} records?`,
```

File: app/src/contexts/telemetryTableContext.tsx [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+            const found = telemetryDataContext.telemetryDataLoader.data?.find(
+              (search) => search.telemetry_manual_id === item.id
+            );
+            if (found) {
+              // existing ID found, update record
+              updateData.push({
+                telemetry_manual_id: String(item.id),
+                latitude: Number(item.latitude),
+                longitude: Number(item.longitude),
+                acquisition_date: moment(moment(item.date).format('YYYY-MM-DD') + ' ' + item.time).format(
```
Comment (curtisupshall) [inline app/src/contexts/telemetryTableContext.tsx]: Nit: use template string

File: app/src/components/loading/SkeletonLoaders.tsx
Comment (al-rosenthal) [reply app/src/components/loading/SkeletonLoaders.tsx:1]: I grabbed those changes and updated the skeleton loads. I also updated the default number of rows to 4

**Review by KjartanE — DISMISSED** [review]
Ran locally, all functions worked properly 🐈

File: app/src/features/surveys/telemetry/ManualTelemetryCard.tsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+      <AccordionDetails
+        sx={{
+          pt: 0,
+          px: 2
+        }}>
+        <Box display={'flex'}>
+          <Icon path={mdiCalendarRange} size={0.75} />
+          <Typography ml={1} variant="body2" component={'div'} color={'inherit'}>
+            {moment(props.start_date).format('YYYY-MM-DD')}{' '}
+            {props.end_date ? '- ' + moment(props.end_date).format('YYYY-MM-DD') : ''}
```
Comment (curtisupshall) [inline app/src/features/surveys/telemetry/ManualTelemetryCard.tsx:71]: Nit: use template string

File: app/src/features/surveys/telemetry/ManualTelemetryList.tsx [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+      </Menu>
+      <Formik
+        initialValues={formData}
+        enableReinitialize
+        validationSchema={AnimalDeploymentSchema}
+        validateOnBlur={false}
+        validateOnChange={true}
+        onSubmit={async (values, actions) => {
+          setIsLoading(true);
+          await handleSubmit(values);
```
Comment (curtisupshall) [inline app/src/features/surveys/telemetry/ManualTelemetryList.tsx:409]: Should probably `.then`, `.catch` this; or use try/catch blocks

File: app/src/features/surveys/view/survey-animals/SurveyAnimalsPage.tsx [hunk 8]
```diff
@@ -99,6 +99,7 @@ export const SurveyAnimalsPage = () => {
       deployments = [];
     }
     animal.device = deployments;
+    console.log(animal);
```
Comment (curtisupshall) [inline app/src/features/surveys/view/survey-animals/SurveyAnimalsPage.tsx]: Nit: remove?

**Review by curtisupshall — COMMENTED** [review]
Tested this and it works well! Just have a handful of minor code comments

File: app/src/features/surveys/telemetry/ManualTelemetryList.tsx (hunk 7 above)
Comment (al-rosenthal) [reply app/src/features/surveys/telemetry/ManualTelemetryList.tsx:409]: the functions called within `handleSubmit` have their own try catches with specific error snack bars

File: app/src/features/surveys/view/survey-animals/SurveyAnimalsPage.tsx (hunk 8 above)
Comment (al-rosenthal) [reply app/src/features/surveys/view/survey-animals/SurveyAnimalsPage.tsx]: removed

File: app/src/features/surveys/telemetry/ManualTelemetryList.tsx (hunk 7 above)
Comment (curtisupshall) [reply app/src/features/surveys/telemetry/ManualTelemetryList.tsx:409]: OK gotcha :)

**Review by al-rosenthal — DISMISSED** [review]
🧑‍🍳

**Review by KjartanE — DISMISSED** [review]
🐈

File: api/src/paths/telemetry/manual/deployments.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
+import { RequestHandler } from 'express';
+import { Operation } from 'express-openapi';
+import { manual_telemetry_responses } from '.';
+import { authorizeRequestHandler } from '../../../request-handlers/security/authorization';
+import { BctwService, getBctwUser } from '../../../services/bctw-service';
+import { getLogger } from '../../../utils/logger';
+
+const defaultLog = getLogger('paths/telemetry/manual');
```
Comment (NickPhura) [inline api/src/paths/telemetry/manual/deployments.ts:8]: super nit: 'paths/telemetry/manual/deployments'

File: api/src/paths/telemetry/manual/deployments.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+export const POST: Operation = [
+  authorizeRequestHandler(() => {
+    return {
+      and: [
+        {
+          discriminator: 'SystemUser'
```
Comment (NickPhura) [inline api/src/paths/telemetry/manual/deployments.ts:15]: Just checking: A bunch of endpoints in this PR specify only SystemUser level security. Is that correct?

File: database/src/migrations/20231204122800_survey_telemetry_submission.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+ * @return {*}  {Promise<void>}
+ */
+export async function up(knex: Knex): Promise<void> {
+  await knex.raw(`--sql
+  ----------------------------------------------------------------------------------------
+  -- Create survey_telemetry_submission table
+  ----------------------------------------------------------------------------------------
+  set search_path=biohub;
+  CREATE TABLE survey_telemetry_submission(
+    submission_id                           integer           NOT NULL,
```
Comment (NickPhura) [inline database/src/migrations/20231204122800_survey_telemetry_submission.ts]: This should be `survey_telemetry_submission_id` to match the current id naming convention.

File: app/src/features/surveys/view/survey-animals/telemetry-device/TelemetryDeviceFormContent.tsx [hunk 12]
```diff
@@ -18,12 +18,25 @@ interface TelemetryDeviceFormContentProps {
...
+      survey_critter_id: '',
+      deployments: [],
+      device_id: '',
+      device_make: '',
+      device_model: '',
+      frequency: '',
+      frequency_unit: ''
+    };
+  }
+  console.log(device);
```
Comment (NickPhura) [inline app/src/features/surveys/view/survey-animals/telemetry-device/TelemetryDeviceFormContent.tsx]: nit: log

File: app/src/features/surveys/telemetry/ManualTelemetryList.tsx [hunk 13]
```diff
@@ -0,0 +1,562 @@
...
+      open: true,
+      onYes: () => {
+        handleDeleteDeployment();
+      }
+    });
+  };
+
+  const handleAddDeployment = async (data: AnimalDeployment) => {
+    const critter = critters?.find((a) => a.survey_critter_id === data.survey_critter_id);
+    if (!critter) console.log('Did not find critter in addTelemetry!');
```
Comment (NickPhura) [inline app/src/features/surveys/telemetry/ManualTelemetryList.tsx]: Is this needed? Ideally we should remove logs like this, especially from the frontend where the public can see them. Also I assume the following code can handle the case where the critter is null?

**Review by al-rosenthal — DISMISSED** [review]
🚗

**Review by NickPhura — APPROVED** [review]

**Review by al-rosenthal — APPROVED** [review]
🥞

**Pull request merged**